File: examples-nosql-go-sdk/sqlexamples/MultiDataOps.go
// Copyright (c) 2023, 2024 Oracle and/or its affiliates.
// Licensed under the Universal Permissive License v 1.0 as shown at
// https://oss.oracle.com/licenses/upl/
package main

import (
	"fmt"
	"time"
	"github.com/oracle/nosql-go-sdk/nosqldb"
	"github.com/oracle/nosql-go-sdk/nosqldb/auth/iam"
	"github.com/oracle/nosql-go-sdk/nosqldb/common"
	"github.com/oracle/nosql-go-sdk/nosqldb/types"
)
// Creates a client with the supplied configurations.
// This function encapsulates environmental differences and returns a
// client handle to use for data operations.
func createClient_cloud() (*nosqldb.Client, error) {
	var cfg nosqldb.Config
	// replace the placeholder with your actual region identifier
	region := "<your_region_identifier>"
	// Replace the placeholders with the actual value of config file location
	// and the ocid of your compartment
	sp, err := iam.NewSignatureProviderFromFile("<location_config_file>", "", "", "<your_compartment_ocid>")
	if err != nil {
		return nil, fmt.Errorf("cannot create a Signature Provider: %v", err)
	}
	cfg = nosqldb.Config{
		Mode:                  "cloud",
		Region:                common.Region(region),
		AuthorizationProvider: sp,
	}
	client, err := nosqldb.NewClient(cfg)
	return client, err
}

// Creates a client with the supplied configurations for onPremise database
func createClient_onPrem() (*nosqldb.Client, error) {
	var cfg nosqldb.Config
	// replace the placeholder with the fullname of your host
	endpoint := "http://<hostname>:8080"
	cfg = nosqldb.Config{
		Endpoint: endpoint,
		Mode:     "onprem",
	}
	// If using a secure store, uncomment the lines below and pass the username,
	// password of the store to Config
	// cfg := nosqldb.Config{
	//    Mode:     "onprem",
	//    Username: "<username>",
	//    Password: []byte("<password>"),
	// Specify InsecureSkipVerify
	//    HTTPConfig: httputil.HTTPConfig{
	//        InsecureSkipVerify: true,
	//    },
	client, err := nosqldb.NewClient(cfg)
	return client, err
}

// Creates a table
func createTable(client *nosqldb.Client, err error, tableName string)(){
	stmt := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s ("+
		"id INTEGER," +
		"address_line1 STRING," +
		"address_line2 STRING," +
		"pin INTEGER ," +
		"PRIMARY KEY(SHARD(pin), id))",tableName)
	tableReq := &nosqldb.TableRequest{
		   Statement: stmt,
		   TableLimits: &nosqldb.TableLimits{
				ReadUnits:  20,
				WriteUnits: 20,
				StorageGB:  1,
			},
	}
	tableRes, err := client.DoTableRequest(tableReq)
	if err != nil {
		fmt.Printf("cannot initiate CREATE TABLE request: %v\n", err)
		return
	}
	// The create table request is asynchronous, wait for table creation to complete.
	_, err = tableRes.WaitForCompletion(client, 60*time.Second, time.Second)
	if err != nil {
		fmt.Printf("Error finishing CREATE TABLE request: %v\n", err)
		return
	}
	fmt.Println("Created table ", tableName)
	return
}
//Add data to table
func insertData(client *nosqldb.Client, err error, tableName string, value1 *types.MapValue )(){
	putReq := &nosqldb.PutRequest{
		TableName: tableName,
		Value:     value1,
	}
	putRes, err := client.Put(putReq)
	if err != nil {
		fmt.Printf("failed to put single row: %v\n", err)
		return
	}
	fmt.Printf("Put row: result: %v\n", putRes)
}
//fetch data from the table
func fetchRowCnt(client *nosqldb.Client, err error, tableName string, querystmt string)(){
	prepReq := &nosqldb.PrepareRequest{
		Statement: querystmt,
	}
   prepRes, err := client.Prepare(prepReq)
	if err != nil {
		fmt.Printf("Prepare failed: %v\n", err)
		return
	}
	queryReq := &nosqldb.QueryRequest{
		 PreparedStatement: &prepRes.PreparedStatement,   }
	var results []*types.MapValue
	for {
		queryRes, err := client.Query(queryReq)
		if err != nil {
			fmt.Printf("Query failed: %v\n", err)
			return
		}
		res, err := queryRes.GetResults()
		if err != nil {
			fmt.Printf("GetResults() failed: %v\n", err)
			return
		}
		results = append(results, res...)
		if queryReq.IsDone() {
			break
		}
	}
	fmt.Printf("Number of query results is : %d\n", len(results))

}
//delete multiple rows
func delMulRows(client *nosqldb.Client, err error, tableName string,pinval int)(){
	shardKey := &types.MapValue{}
	shardKey.Put("pin", pinval)
	multiDelReq := &nosqldb.MultiDeleteRequest{
		TableName: tableName,
		Key:       shardKey,
	}
	multiDelRes, err := client.MultiDelete(multiDelReq)
	if err != nil {
		fmt.Printf("failed to delete multiple rows: %v", err)
		return
	}
	fmt.Printf("MultiDelete result=%v\n", multiDelRes)
}


func main() {
	//if using cloud service uncomment the line below
	client, err := createClient_cloud()
	//if using onPrem uncomment the line below
	//client, err := createClient_onPrem()
	 if err != nil {
		fmt.Printf("cannot create NoSQL client: %v\n", err)
		return
	}
	defer client.Close()
	tableName := "examplesAddress"
	createTable(client, err,tableName)
	//adding data
	value, err := types.NewMapValueFromJSON(`{
		"id":            1,
		"pin":           1234567,
		"address_line1": "10 Red Street",
		"address_line2": "Apt 3"
	}`)
	insertData(client, err,tableName,value)
	value1, err := types.NewMapValueFromJSON(`{
		"id":            2,
		"pin":           1234567,
		"address_line1": "2 Green Street",
		"address_line2": "Street 9"
	}`)
	insertData(client, err,tableName,value1)
	value2, err := types.NewMapValueFromJSON(`{
		"id":            3,
		"pin":           1234567,
		"address_line1": "5 Blue Ave",
		"address_line2": "Floor 2"
	}`)
	insertData(client, err,tableName,value2)
	value3, err := types.NewMapValueFromJSON(`{
		"id":            4,
		"pin":           87654321,
		"address_line1": "9 Yellow Boulevard",
		"address_line2": "Apt 3"
	}`)
	insertData(client, err,tableName,value3)
	fmt.Printf("Put row succeeded: \n")
	stmt1 := "select * from examplesAddress"
	fetchRowCnt(client, err,tableName,stmt1)
	delMulRows(client, err,tableName,1234567)
	fetchRowCnt(client, err,tableName,stmt1)
}

File: examples-nosql-node-sdk/sqlexamples/typescript/MultiDataOps.ts
/* Copyright (c) 2023, 2024 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at
 * https://oss.oracle.com/licenses/upl/
 */
'use strict';
import { NoSQLClient, Region } from 'oracle-nosqldb'; //Uncomment for Cloud
/* import { NoSQLClient, ServiceType  } from 'oracle-nosqldb'; */ //Uncomment for On-Premise

interface AddressInt {
   id: Integer;
   address_line1: String;
   address_line2: String;
   pin: Integer;
}
const TABLE_NAME = 'examplesAddress';
const add1= `
{
   "id":1,
   "address_line1":"10 Red Street",
   "address_line2":"Apt 3",
   "pin":1234567
}`
const add2= `
{
   "id":2,
   "address_line1":"2 Green Street",
   "address_line2":"Street 9",
   "pin":1234567
}`
const add3= `
{
   "id":3,
   "address_line1":"5 Blue Ave",
   "address_line2":"Floor 9",
   "pin":1234567
}`
const add4= `
{
   "id":4,
   "address_line1":"9 Yellow Boulevard",
   "address_line2":"Apt 3",
   "pin":87654321
}`
const stmt1 = 'select * from examplesAddress';


/**
  * Call the main function 
  **/
doMultiDataOps();

async function doMultiDataOps() {
   let handle;
   try {
      /* UNCOMMENT line of code below if you are using Oracle NoSQL Database
      * Cloud service. Leave the line commented if you are using onPrem database*/
      handle = await getConnection_cloud();
      /* UNCOMMENT line of code below if you are using onPremise Oracle NoSQL
       * Database. Leave the line commented if you are using NoSQL Database
       * Cloud Service
      handle = await getConnection_onPrem(); */
      await createTable(handle);
      let putResult = await handle.put<AddressInt>(TABLE_NAME, JSON.parse(add1));
      let putResult1 = await handle.put<AddressInt>(TABLE_NAME, JSON.parse(add2));
      let putResult2 = await handle.put<AddressInt>(TABLE_NAME, JSON.parse(add3));
      let putResult3 = await handle.put<AddressInt>(TABLE_NAME, JSON.parse(add4));
      console.log("Wrote records of examplesAddress schema");
      await fetchRowCnt(handle,stmt1);
      await mulRowDel(handle,1234567);
      await fetchRowCnt(handle,stmt1);
   } catch (error ) {
      console.log(error);
   }
   finally {
      if (handle) {
         handle.close();
      }
   }
}

/* Create and return an instance of a NoSQLCLient object for cloud service */
function getConnection_cloud() {
   /* replace the placeholders for compartment and region with actual values.*/
   const Region = `<your_region_identifier>`;
   return new NoSQLClient({
      region: Region,
      compartment: "<ocid_of_your_compartment>",
   });
}
/* Create and return an instance of a NoSQLCLient object for onPremises */
function getConnection_onPrem() {
   /* replace the placeholder with the fullname of your host */
   const kvstore_endpoint = `http://<hostname>:8080`;
   return new NoSQLClient({
      serviceType: "KVSTORE",
      endpoint: kvstore_endpoint
   });
   /* if it is a secure store, comment the return statement above and
   * uncomment the lines below and pass the credentials, username and password
   return new NoSQLClient({
      "serviceType": "KVSTORE",
      "endpoint": "",
      "auth": {
         "kvstore":{
            "user": "",
            "password": ""
         }
      }
   }); */
}

//creates a table
async function createTable(handle: NoSQLClient) {
   const createDDL = `CREATE TABLE IF NOT EXISTS ${TABLE_NAME} (id INTEGER,
                                                                 address_line1 STRING,
                                                                 address_line2 STRING,
                                                                 pin INTEGER,
                                                                 PRIMARY KEY(SHARD(pin), id))`;
   console.log('Create table: ' + createDDL);
   let res =  await handle.tableDDL(createDDL, {
            complete: true,
            tableLimits: {
                readUnits: 20,
                writeUnits: 20,
                storageGB: 1
            }
   });
   console.log('Table created: ' + TABLE_NAME);
}
//fetches data from the table
async function fetchRowCnt(handle: NoSQLClient,querystmt: string) {
   const opt = {};
   try {
      do {
         const result = await handle.query<AddressInt>(querystmt, opt);
         for(let row of result.rows) {
            console.log('  %O', row);
         }
         opt.continuationKey = result.continuationKey;
      } while(opt.continuationKey);
   } catch(error) {
      console.error('  Error: ' + error.message);
   }
}

//deletes multiple rows
async function mulRowDel(handle: NoSQLClient,pinval: string){
   try {
      /* Unconditional delete, should succeed.*/
      var result = await handle.deleteRange<AddressInt>(TABLE_NAME, { pin: pinval });
      /* Expected output: delete succeeded*/
      console.log('delete ' + result.success ? 'succeeded' : 'failed');
   } catch(error) {
      console.error('  Error: ' + error.message);
   }
}
